### PR TITLE
DBZ-5410: Ability to read commit log files incrementally in Cassandra4

### PR DESCRIPTION
**Pull request by nitinitt**
…4 and not when its marked Complete

### CODE REVIEW

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java [hunk 1]
```diff
@@ -246,38 +254,67 @@ private ProcessingResult callInternal() {
...
+                    if ((segmentToOffset.get(commitLog.commitLogSegmentId) == null)) {
+                        LOGGER.info("Start to read the partial file : {}", commitLog.toString());
+                        commitLogPosition = new CommitLogPosition(commitLog.commitLogSegmentId, 0);
+                    }
+                    else if (segmentToOffset.get(commitLog.commitLogSegmentId) < commitLog.offsetOfEndOfLastWrittenCDCMutation) {
+                        LOGGER.info("Resume to read the partial file : {}", commitLog.toString());
+                        commitLogPosition = new CommitLogPosition(commitLog.commitLogSegmentId, segmentToOffset.get(commitLog.commitLogSegmentId));
+                    }
+                    else {
+                        LOGGER.info("No movement in HWM in IDX file: {}", commitLog.toString());
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java:274]: what does HWM stands for?

Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java:274]: HWM stands for High Watermark, I understand and will update the log statement as HWM is not appropriate

File: core/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java [hunk 2]
```diff
@@ -150,6 +150,19 @@ public static Optional<SnapshotMode> fromText(String text) {
...
             .withDescription("Determines whether or not the CommitLogProcessor should re-process error commitLogFiles.");
 
+    /**
+     * Only valid for Cassandra4 and if enabled, commit logs would be read incrementally instead of reading complete log file
+     */
+    public static final boolean DEFAULT_COMMIT_LOG_REAL_TIME_PROCESSING_ENABLED = false;
+    public static final Field COMMIT_LOG_REAL_TIME_PROCESSING_ENABLED = Field.create("commit.log.real.time.processing.enabled")
+            .withType(Type.BOOLEAN)
+            .withDefault(DEFAULT_COMMIT_LOG_REAL_TIME_PROCESSING_ENABLED)
+            .withDescription("Enables the near real time processing of commit logs for Cassandra4 by reading commit log files incrementally");
```
Comment (smiklosovic) [inline core/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java:160]: Cassandra 4

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java [hunk 3]
```diff
@@ -96,16 +94,18 @@ public void stop() {
...
     }
 
-    final static Set<Pair<CommitLogProcessingCallable, Future<ProcessingResult>>> submittedProcessings = ConcurrentHashMap.newKeySet();
-
     private void submit(Path index) {
-        CommitLogProcessingCallable callable = new CommitLogProcessingCallable(new LogicalCommitLog(index.toFile()),
-                queues,
-                metrics,
-                Cassandra4CommitLogProcessor.this.context);
+        final Cassandra4CommitLogParserBase cassandra4CommitLogParserBase;
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java:98]: it should be called just `parser`.

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java [hunk 4]
```diff
@@ -96,16 +94,18 @@ public void stop() {
...
 
-        Future<ProcessingResult> future = executorService.submit(callable);
-        submittedProcessings.add(new Pair<>(callable, future));
+        if (context.getCassandraConnectorConfig().isCommitLogRealTimeProcessingEnabled()) {
+            cassandra4CommitLogParserBase = new Cassandra4CommitLogNearRealTimeParser(new LogicalCommitLog(index.toFile()), queues, metrics, this.context);
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java:101]: you could simplify this, both constructors are taking same variables in.

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogNearRealTimeParser.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.List;
+
+import org.apache.cassandra.db.commitlog.CommitLogPosition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.connector.base.ChangeEventQueue;
+
+public class Cassandra4CommitLogNearRealTimeParser extends Cassandra4CommitLogParserBase {
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogNearRealTimeParser.java:16]: I dont like the name of this class, it is little bit .... verbose. Not sure what to call it like though.

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogParserBase.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import org.apache.cassandra.db.commitlog.CommitLogPosition;
+import org.apache.cassandra.db.commitlog.CommitLogReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.DebeziumException;
+import io.debezium.connector.base.ChangeEventQueue;
+import io.debezium.connector.cassandra.exceptions.CassandraConnectorTaskException;
+
+public abstract class Cassandra4CommitLogParserBase {
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogParserBase.java:21]: I would rather call this `AbstractCassandra4CommitLogParser`

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        try {
+            parseIndexFile(commitLog);
+
+            while (!commitLog.completed) {
+                if (completePrematurely) {
+                    LOGGER.info("{} completed prematurely", commitLog.toString());
+                    return new Cassandra4CommitLogProcessor.ProcessingResult(commitLog, Cassandra4CommitLogProcessor.ProcessingResult.Result.COMPLETED_PREMATURELY);
+                }
+                // TODO make this configurable maybe
+                Thread.sleep(10000);
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java:36]: would you mind to make this configurable as part of this PR?

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public Cassandra4CommitLogProcessor.ProcessingResult parse() {
+        try {
+            parseIndexFile(commitLog);
+
+            while (!commitLog.completed) {
+                if (completePrematurely) {
+                    LOGGER.info("{} completed prematurely", commitLog.toString());
+                    return new Cassandra4CommitLogProcessor.ProcessingResult(commitLog, Cassandra4CommitLogProcessor.ProcessingResult.Result.COMPLETED_PREMATURELY);
```
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java:33]: could not you make `COMPLETED_PREMATURELY` as import?

File: core/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java (hunk 2 above)
Comment (nitinitt) [reply core/src/main/java/io/debezium/connector/cassandra/CassandraConnectorConfig.java]: done

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java (hunk 3 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java]: done

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogNearRealTimeParser.java (hunk 5 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogNearRealTimeParser.java]: Dropped the word "near" from the name, hope it helps to shorten it.

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogParserBase.java (hunk 6 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogParserBase.java]: done

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java (hunk 7 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java]: done

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java (hunk 8 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogBatchParser.java]: done

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java (hunk 1 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java]: addressed

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java (hunk 4 above)
Comment (nitinitt) [reply cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogProcessor.java]: Sorry, I did not understand here. Do you want to create a method in abstract class and set member variables at a common place, and that setter is called outside of if/else block?

**Comment by nitinitt** [comment]
Hey @smiklosovic,
I have incorporated the code review feedback, please let me know if the code looks good.

**Comment by jpechane** [comment]
Continues in https://github.com/debezium/debezium-connector-cassandra/pull/62

**Comment by smiklosovic** [comment]
@jpechane I think you meant #69

**Comment by jpechane** [comment]
@smiklosovic Yes, thanks for the notification!

**Pull request closed without merging**